### PR TITLE
Properly handle empty list of role requests

### LINKED ISSUE #13411: Machine ID: more gracefully handle lack of bot roles
If a bot user has not been configured with any roles, and the user hasn't configured any roles for `tbot` to request, when the bot attempts to generate impersonated certificates, it will instead receive normal bot certificates (as no roles will be requested by the bot). This will cause the next renewal to cause a generation mismatch lockout.

### PR DESCRIPTION
**Pull request by timothyb89**
Currently, an empty list of role requests results in an ambiguous situation: we usually use the presence of role requests to determine whether or not we'd return impersonated certs or not. An empty list of role requests returns a fresh set of non-impersonated certs (possibly renewed if allowed), while a non-empty list of role requests returns certs with just those roles.  However, if a client _intends_ to request impersonated certs but provides an empty list of role requests it will instead receive non-impersonated (possibly renewable) certs with the full permissions of the original user.

This could theoretically result in privilege escalation if a Machine ID bot: (a) had any worthwhile permissions of its own, which is not the case unless the bot role was manually modified and (b) accidentally handed certs off to an attacker.

In practice this bug is fairly difficult to hit: `tbot` always auto-fills all requestable roles if they are otherwise unset, and `tctl bots add` requires `--roles=` to be passed. An empty string here can trigger the bug however it is unlikely a user would pass this by accident. Moreover, a bot without requestable roles cannot accomplish much of anything, so this is exceedingly unlikely to be intended behavior.

Additionally, certificate generation checks help to mitigate the issue: bots currently lock themselves by accident after the first renewal when this bug is triggered as they don't explicitly handle receiving renewable certs when impersonated certs are expected. If an attacker were to attempt a renewal, the generation counter would similarly limit access, and as noted previously, the bot role grants only minimal read-only access anyway.

---

To resolve the issue, this adds a new `RoleRequestsOnly` flag to `UserCertsRequest` that allows clients to unambiguously specify if they wish to receive a non-impersonated, possibly renewable, cert with all the original user's roles and permissions, or if they wish to receive only role-impersonated certs (or an error if roles are empty). Machine ID passes this flag in all situations where an impersonated cert is desired.

Additionally, we also now ensure users add at least one role in `CreateBot()` (called by `tctl bots add`) as this is almost certainly an unintended situation.

Fixes #13411

### CODE REVIEW

**Comment by strideynet** [comment]
Thanks for raising this. Nice to get this bug closed up. I'll take a look at this PR a little later.

File: lib/auth/auth_with_roles_test.go [hunk 1]
```diff
@@ -258,6 +266,18 @@ func TestBotCertificateGenerationStolen(t *testing.T) {
...
+// TestBotNoRoles attempts to create a bot with an empty role list.
+func TestBotNoRoles(t *testing.T) {
+	t.Parallel()
+	srv := newTestTLSServer(t)
+
+	// Create a new bot.
+	_, err := srv.Auth().createBot(context.Background(), &proto.CreateBotRequest{
+		Name: "test",
+	})
+	require.Error(t, err)
```
Comment (strideynet) [inline lib/auth/auth_with_roles_test.go:278]: nit: It might be nice to look for a specific error here.

File: lib/auth/auth_with_roles.go [hunk 2]
```diff
@@ -1924,9 +1924,15 @@ func (a *ServerWithRoles) NewKeepAliver(ctx context.Context) (types.KeepAliver,
 // user is allowed to assume the returned roles.
 func (a *ServerWithRoles) determineDesiredRolesAndTraits(req proto.UserCertsRequest, user types.User) ([]string, wrappers.Traits, error) {
 	if req.Username == a.context.User.GetName() {
+		// If RoleRequestsOnly is set, make sure we don't return unusable
+		// certs: an identity without roles can't be parsed.
+		if req.RoleRequestsOnly && len(req.RoleRequests) == 0 {
+			return nil, nil, trace.BadParameter("at least one role request is required")
+		}
```
Comment (strideynet) [inline lib/auth/auth_with_roles.go:1931]: Do we want to add a future version date in which we will deprecate the ability to provide `req.RoleRequests` without the `RoleRequestsOnly` flag ?

**Review by strideynet — APPROVED** [review]
Looks good to me, just two thoughts.

**Comment by timothyb89** [comment]
> > tctl bots add requires --roles= to be passed. An empty string here can trigger the bug however it is unlikely a user would pass this by accident.
> 
> Is an empty string for `--roles` a valid input? Should this PR also add update `tctl` to prevent this if it isn't valid?

This PR also makes the `CreateBot()` gRPC call fail, so you'll now see this:

```
$ tctl bots add test --roles=
ERROR: error while creating bot
cannot add bot: at least one role is required
```

I figured it would be best to make it as exhaustive as possible.

File: lib/auth/auth_with_roles.go (hunk 2 above)
Comment (timothyb89) [reply lib/auth/auth_with_roles.go:1933]: That's probably for the best. I've added a note and preemptively renamed the parameter to `UseRoleRequests` which I think will make more sense when it's required.

**Review by rosstimothy — APPROVED** [review]
I still feel like a check in `tctl bots add` to ensure that `len(roles) > 0` would be warranted in addition to the server side check. If anything, it's one less wasted trip to the auth server. There are other `tctl` commands that verify input and fail if things are not provided so this wouldn't be unprecedented behavior either.

**Comment by timothyb89** [comment]
> I still feel like a check in `tctl bots add` to ensure that `len(roles) > 0` would be warranted in addition to the server side check. If anything, it's one less wasted trip to the auth server. There are other `tctl` commands that verify input and fail if things are not provided so this wouldn't be unprecedented behavior either.

Fair enough! I've added a local check as well.

**Pull request merged**